task/queue/queue.go

```diff
@@ -0,0 +1,80 @@
+package queue
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+
+	"github.com/panjf2000/ants/v2"
+)
+
+// TaskQueue representation of task queue.
+type TaskQueue struct {
+	goPool             *ants.Pool
+	limit              int
+	numParallelRequest int
+
+	mutex sync.Mutex
+	queue *list.List
+}
+
+// New creates task queue which is processed by number of workers. Number of task is limited by limit.
+func New(numWorkers, limit int) (*TaskQueue, error) {
+	p, err := ants.NewPool(numWorkers, ants.WithPreAlloc(true), ants.WithNonblocking(true))
+	if err != nil {
+		return nil, err
+	}
+	return &TaskQueue{
+		queue:              list.New(),
+		goPool:             p,
+		limit:              limit,
+		numParallelRequest: numWorkers,
+	}, nil
+}
+
+func (q *TaskQueue) appendQueue(tasks []func()) error {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
```

Comment:
Why are you locking here and pre-checking the size of ants queue? This will be done when submitting anyway, so error would come from there.

Reply:
No this is not possible. Ants doesn't support queue of tasks.
Locking is needed because `q.queue` is not thread safe